pkg/controller/storagecluster/storagecluster.go

```diff
@@ -643,6 +640,108 @@ func (c *Controller) syncNodes(
 	return utilerrors.NewAggregate(errors)
 }
 
+func (c *Controller) podTemplatesForNodes(
+	cluster *corev1alpha1.StorageCluster,
+	nodesNeedingStoragePods []string,
+	hash string,
+) ([]string, []*v1.PodTemplateSpec, error) {
+	nodeList := &v1.NodeList{}
+	err := c.client.List(context.TODO(), nodeList, &client.ListOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	remainingNodes := make(map[string]*v1.Node)
+	for _, nodeName := range nodesNeedingStoragePods {
+		for _, node := range nodeList.Items {
+			if node.Name == nodeName {
+				remainingNodes[nodeName] = node.DeepCopy()
+				break
+			}
+		}
+	}
+
+	podTemplates := make([]*v1.PodTemplateSpec, 0)
+	nodesNeedingStoragePods = make([]string, 0)
+
+	for _, nodeSpec := range cluster.Spec.Nodes {
+		if len(remainingNodes) == 0 {
+			break
+		}
+
+		nodeGroup := make([]*v1.Node, 0)
+		if nodeSpec.Selector.NodeName != "" {
+			if node, exists := remainingNodes[nodeSpec.Selector.NodeName]; exists {
+				nodeGroup = append(nodeGroup, node)
+			} else {
+				continue
+			}
+		} else {
+			nodeSelector, err := metav1.LabelSelectorAsSelector(nodeSpec.Selector.LabelSelector)
+			if err != nil {
+				logrus.Warnf("Failed to parse label selector %#v: %v", nodeSpec.Selector.LabelSelector, err)
+				continue
+			}
+			for _, node := range remainingNodes {
+				if nodeSelector.Matches(labels.Set(node.Labels)) {
+					nodeGroup = append(nodeGroup, node)
+				}
+			}
+		}
+
+		if len(nodeGroup) > 0 {
+			clusterForNodeGroup := cluster.DeepCopy()
+			updateClusterSpecWithNodeSpec(&clusterForNodeGroup.Spec, nodeSpec.DeepCopy())
+
+			if err := c.createPodTemplateForNodeGroup(
+				clusterForNodeGroup, nodeGroup,
+				&nodesNeedingStoragePods, &podTemplates,
+				remainingNodes, hash,
+			); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+
+	if len(remainingNodes) > 0 {
+		nodeGroup := make([]*v1.Node, 0)
+		for _, node := range remainingNodes {
+			nodeGroup = append(nodeGroup, node)
+		}
+
+		if err := c.createPodTemplateForNodeGroup(
+			cluster, nodeGroup,
+			&nodesNeedingStoragePods, &podTemplates,
+			remainingNodes, hash,
+		); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return nodesNeedingStoragePods, podTemplates, nil
+}
+
+func (c *Controller) createPodTemplateForNodeGroup(
+	cluster *corev1alpha1.StorageCluster,
+	nodeGroup []*v1.Node,
+	nodesNeedingStoragePods *[]string,
+	podTemplates *[]*v1.PodTemplateSpec,
+	remainingNodes map[string]*v1.Node,
+	hash string,
+) error {
+	podTemplate, err := c.createPodTemplate(cluster, nodeGroup[0], hash)
```

Comment:
Shouldn't this be inside the for loop ?
Why use the 1st node for all pod templates?

Reply:
All the nodes in nodeGroup will have the same template, so no need to call createPodTemplate for each one of the nodes. That is why we are grouping them.
In most cases (without any node spec) there is going to be a single template for all pods, so wanted to optimize this.